pallets/settlement/src/lib.rs

```diff
@@ -1209,74 +1205,90 @@ impl<T: Config> Module<T> {
+                    instruction_asset_count.off_chain(),
+                ))
+                .map_err(|_| Error::<T>::WeightLimitExceeded)?;
 
-        // Verifies if the venue is allowed for all tickers in the instruction
-        Self::ensure_allowed_venue(&instruction_legs, venue_id)?;
+            // Ensures the venue is allowed for all tickers in the instruction
+            let instruction_details = InstructionDetails::<T>::take(instruction_id);
+            Self::ensure_allowed_venue(&instruction_legs, instruction_details.venue_id)?;
 
-        let instruction_memo = InstructionMemos::get(&instruction_id);
-        // Attempts to release the locks and transfer all fungible an non fungible assets
-        if let Err(leg_id) = frame_storage_with_transaction(|| {
-            Self::release_asset_locks_and_transfer_pending_legs(
+            // Attempts to release the locks and transfer all fungible an non fungible assets
+            let instruction_memo = InstructionMemos::get(&instruction_id);
+            match Self::release_asset_locks_and_transfer_pending_legs(
                 instruction_id,
                 &instruction_legs,
                 instruction_memo,
                 caller_did,
                 weight_meter,
-            )
-        })? {
-            Self::deposit_event(RawEvent::LegFailedExecution(
-                caller_did,
-                instruction_id,
-                leg_id,
-            ));
-            Self::deposit_event(RawEvent::InstructionFailed(caller_did, instruction_id));
-            return Err(Error::<T>::FailedToReleaseLockOrTransferAssets.into());
-        }
-
-        Self::deposit_event(RawEvent::InstructionExecuted(caller_did, instruction_id));
-        Ok(())
+            ) {
+                Ok(_) => {
+                    // Remove remaning storage
+                    VenueInstructions::remove(instruction_details.venue_id, instruction_id);
+                    let _ = InstructionLegStatus::<T>::clear_prefix(
+                        instruction_id,
+                        instruction_legs.len() as u32,
+                        None,
+                    );
+                    // Change instruction status
+                    InstructionStatuses::<T>::insert(
+                        instruction_id,
+                        InstructionStatus::Success(System::<T>::block_number()),
+                    );
+                    Self::deposit_event(RawEvent::InstructionExecuted(caller_did, instruction_id));
+                    Ok(())
+                }
+                Err(leg_id) => {
+                    Self::deposit_event(RawEvent::LegFailedExecution(
+                        caller_did,
+                        instruction_id,
+                        leg_id,
+                    ));
+                    Self::deposit_event(RawEvent::InstructionFailed(caller_did, instruction_id));
+                    Err(Error::<T>::FailedToReleaseLockOrTransferAssets.into())
+                }
```

Comment:
This is still inside the `with_transaction`, so it would be reverted too and those events would be lost.